### PR TITLE
Week2

### PR DESCRIPTION
**Pull request by mikhailgladchenko**
solutions for week2 tasks

### CODE REVIEW

File: week2/ex4_done.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    def __cmp__(self, other):
+        return User.cmp(self.age, other.age)
+
+
+def main():
+    users = [User("mikhail", 57), User("marina", 55), User("alexandra", 19), User("mikhail", 31)]
+    print(users)
+    sorted_users_by_age = sorted(users)
+    print(sorted_users_by_age)
+
```
Comment (frizzby) [inline week2/ex4_done.py]:
```python
print(User("mikhail", 57) == User("mikhail", 57))
```
Check this test case.

File: week2/ex1_done.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    new_url = ""
+
+    for key in keys_sorted:
+
+        new_url += "{}={}{}".format(key, params_dict[key], "&")
+
+    return new_url[0:-1]
+
+
+def main():
```
Comment (frizzby) [inline week2/ex1_done.py]: Please add more test cases. Test corner cases (like empty string, etc) as well.

File: week2/ex2_done.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    new_url = ""
+
+    for key in keys_original:
+
+        new_url += "{}={}{}".format(key, params_dict[key], "&")
+
+    return new_url[0:-1]
+
+
+def main():
```
Comment (frizzby) [inline week2/ex2_done.py]: Please add more test cases.

File: week2/ex3_done.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    """
+    Overrides the 5 methods of a MutableMapping:
+    __getitem__, __setitem__, __delitem__, __iter__, __len__
+
+    ...and the 3 non-Mapping methods of Counter:
+    elements, most_common, subtract
+    """
+
+    def __init__(self, values=None, *, mutator=_identity):
+        self._mutator = mutator
```
Comment (frizzby) [inline week2/ex3_done.py]: What is `mutator`? Why you need it here?

File: week2/ex3_done.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        if values is None:
+            self._counter = collections.Counter()
```
Comment (frizzby) [inline week2/ex3_done.py:19]: The tasks asks to implement your own Counter class. Not to wrap around the standard one. 
Your code should have no references to `collections.Counter`

**Review by frizzby — CHANGES_REQUESTED** [review]

File: week2/ex4_done.py (hunk 1 above)
Comment (frizzby) [reply week2/ex4_done.py]: what about this? Did you check it?

File: week2/ex3_mycounter_done.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        return repr(self.mydict)
+
+    def most_common(self, n):
+        counter = 0
+        mc = {}
+        for key, value in sorted(self.mydict.items(), key=lambda kv: kv[1], reverse=True):
+            if counter < n:
+                mc[key] = value
+            counter += 1
+        return mc
```
Comment (frizzby) [inline week2/ex3_mycounter_done.py]: This has not the same return value as in [`collections.Counter`](https://docs.python.org/3/library/collections.html#collections.Counter.most_common). Please make it equivalent.

File: week2/ex3_mycounter_done.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            self.mydict = {}
+
+    def __repr__(self):
+        return repr(self.mydict)
+
+    def most_common(self, n):
+        counter = 0
+        mc = {}
+        for key, value in sorted(self.mydict.items(), key=lambda kv: kv[1], reverse=True):
+            if counter < n:
```
Comment (frizzby) [inline week2/ex3_mycounter_done.py]: Instead of counting n items, you can just slice n first elements: `sorted(...)[:n]`

File: week2/ex3_done.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
+# 3.	Elements counting. It refers to the sections 1.12 and 1.4-1.5 of the Python Cookbook. Implement the Counter class. It should only support operations like __iter__, __getitem__ and .most_common(n). Try to make it as efficient as possible in terms of performance. Think of an underlying data structure that should be used and also a base class.
```
Comment (frizzby) [inline week2/ex3_done.py:1]: Please remove this file.

File: week2/ex2_done.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+    for key in keys_original:
+
+        new_url += "{}={}{}".format(key, params_dict[key], "&")
+
+    return None if new_url == str("=&") else new_url[0:-1]
+
+
+def main():
+    url = "page=2&foo=bar&x=y&utm_source=ss&utm_content=ccc"
```
Comment (frizzby) [inline week2/ex2_done.py]: This code has a few flaws.
Add more test cases and you'll see it.

This code does not retain the original params order.

File: week2/ex1_done.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
+# 1.	Write a function that will parse a url query string like "page=2&foo=bar&x=y&utm_source=ss&utm_content=ccc" and return a query string where all the params are sorted by their names like this "foo=bar&page=2&x=y" and all the "utm_" parameters are removed.
```
Comment (frizzby) [inline week2/ex1_done.py:1]: _all the "utm\_" parameters are removed_ here means the all parameters that START with `utm_`. But your code removes all that contain _utm\__ substring.

File: week2/ex1_done.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+        except IndexError:
+            pass
+        if "utm_" not in key:
+            params_dict[key] = value
+    keys_sorted = sorted(params_dict.keys())
+    new_url = ""
+
+    for key in keys_sorted:
+
+        new_url += "{}={}{}".format(key, params_dict[key], "&")
```
Comment (frizzby) [inline week2/ex1_done.py]: Don't use `+=` or `+` to build strings as it is very inefficient. Strings in python are immutable, and each concatenation creates a new string which is slow and takes memory. And also it hurts readability. 

Please use [`str.join()`](https://docs.python.org/3/library/stdtypes.html#str.join) instead.

File: week2/ex1_done.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
+# 1.	Write a function that will parse a url query string like "page=2&foo=bar&x=y&utm_source=ss&utm_content=ccc" and return a query string where all the params are sorted by their names like this "foo=bar&page=2&x=y" and all the "utm_" parameters are removed.
+
+
+def parse_url(url):
+    string_splitted = url.split("&")
+    params_dict = {}
+    for entry in string_splitted:
+        key = value = ""
+        try:
+            key = entry.split("=")[0]
```
Comment (frizzby) [inline week2/ex1_done.py]: `key, value = entry.split()`

**Review by frizzby — CHANGES_REQUESTED** [review]

File: week2/ex1_done.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
+# 1.	Write a function that will parse a url query string like "page=2&foo=bar&x=y&utm_source=ss&utm_content=ccc" and return a query string where all the params are sorted by their names like this "foo=bar&page=2&x=y" and all the "utm_" parameters are removed.
+
+
+def parse_url(url):
+    string_splitted = url.split("&")
+    params_dict = {}
+    for entry in string_splitted:
```
Comment (frizzby) [inline week2/ex1_done.py]: What you are doing here is basically filtering a list. This is very common operation in Python and it is done best with [list comprehensions](https://docs.python.org/3.6/tutorial/datastructures.html#list-comprehensions):

```python
filtered = [x for x in string_splitted if not x.startswith("utm_")]
```

Another approach is to use [`filter()`](http://book.pythontips.com/en/latest/map_filter.html):
```python
filtered = filter(lambda x: not x.startswith("utm_"), string_splitted)
```


The basic operations on lists are: `filter()`, `map()`, `reduce()` ([read](http://book.pythontips.com/en/latest/map_filter.html)). Keep them in mind when dealing with lists. You can solve almost any task just with this three functions.

**Comment by mikhailgladchenko** [comment]
Yan,
I’ve just pushed new code for ex3_mycounter_done with different implementations of most_common function. I have there the one that slices the list (most_common) and another one that uses heapq.nlargest api(most_common_heap).I’m not sure that the last one works as it should.Can we take a look togather on Monday?
Regards,Mikhail
